compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/MemoizeFbtAndMacroOperandsInSameScope.ts

```diff
@@ -47,245 +100,208 @@ import {Iterable_some} from '../Utils/utils';
       }
     }
   }
+  return macroTags;
 }
 
-function mergeScopes(
-  root: Identifier,
-  scope: ReactiveScope,
-  macroValues: Map<Identifier, Array<Identifier>>,
-  macroTagsCalls: Set<IdentifierId>,
-): void {
-  const operands = macroValues.get(root);
-  if (operands == null) {
-    return;
-  }
-  for (const operand of operands) {
-    operand.scope = scope;
-    expandFbtScopeRange(scope.range, operand.mutableRange);
-    macroTagsCalls.add(operand.id);
-    mergeScopes(operand, scope, macroValues, macroTagsCalls);
-  }
-}
-
-function matchesExactTag(s: string, tags: Set<Macro>): boolean {
-  return Array.from(tags).some(macro =>
-    typeof macro === 'string'
-      ? s === macro
-      : macro[1].length === 0 && macro[0] === s,
-  );
-}
-
-function matchTagRoot(
-  s: string,
-  tags: Set<Macro>,
-): Array<Array<MacroMethod>> | null {
-  const methods: Array<Array<MacroMethod>> = [];
-  for (const macro of tags) {
-    if (typeof macro === 'string') {
-      continue;
+function mergeMacroArguments(
+  fn: HIRFunction,
+  macroTags: Map<IdentifierId, MacroDefinition>,
+  macroKinds: Map<Macro, MacroDefinition>,
+): Set<IdentifierId> {
+  const macroValues = new Set<IdentifierId>(macroTags.keys());
+  for (const block of Array.from(fn.body.blocks.values()).reverse()) {
+    for (let i = block.instructions.length - 1; i >= 0; i--) {
+      const instr = block.instructions[i]!;
+      const {lvalue, value} = instr;
+      switch (value.kind) {
+        case 'CallExpression':
+        case 'MethodCall': {
+          const scope = lvalue.identifier.scope;
+          if (scope == null) {
+            continue;
+          }
+          const callee =
+            value.kind === 'CallExpression' ? value.callee : value.property;
+          const macroDefinition =
+            macroTags.get(lvalue.identifier.id) ??
+            macroTags.get(callee.identifier.id);
+          if (macroDefinition != null) {
+            macroValues.add(lvalue.identifier.id);
+            for (const operand of eachInstructionValueOperand(value)) {
+              macroValues.add(operand.identifier.id);
+              operand.identifier.scope = scope;
+              expandFbtScopeRange(scope.range, operand.identifier.mutableRange);
+              if (macroDefinition.level === InlineLevel.Transitive) {
+                macroTags.set(operand.identifier.id, macroDefinition);
+              }
+            }
+          }
+          break;
+        }
+        case 'JsxExpression': {
+          const scope = lvalue.identifier.scope;
+          if (scope == null) {
+            continue;
+          }
+          let macroDefinition = macroTags.get(lvalue.identifier.id);
+          if (macroDefinition == null) {
+            if (value.tag.kind === 'Identifier') {
+              macroDefinition = macroTags.get(value.tag.identifier.id);
+            } else {
+              macroDefinition = macroKinds.get(value.tag.name);
+            }
+          }
```

Comment:
note that the transitive/shallow distinction is currently meaningless because of the PromoteUsedTemporaries issue - to fix it we have to either allow reordering of macro-related temporaries or force the inlining to be transitive, even for things like fbt.plural that don't strictly require their arguments to be transitively inlined